core/unix/signal.c

```diff
@@ -2293,20 +2293,43 @@ check_signals_pending(dcontext_t *dcontext, thread_sig_info_t *info)
 /* Returns whether to execute the syscall */
 bool
 handle_sigprocmask(dcontext_t *dcontext, int how, kernel_sigset_t *app_set,
-                   kernel_sigset_t *oset, size_t sigsetsize, uint *errno_val)
+                   kernel_sigset_t *oset, size_t sigsetsize, uint *error_code)
 {
     thread_sig_info_t *info = (thread_sig_info_t *)dcontext->signal_field;
     int i;
     kernel_sigset_t safe_set, safe_old_set;
-    if ((app_set != NULL && !d_r_safe_read(app_set, sizeof(safe_set), &safe_set)) ||
-        (oset != NULL && !d_r_safe_read(oset, sizeof(safe_old_set), &safe_old_set))) {
-        if (errno_val != NULL)
-            *errno_val = EFAULT;
+    /* Some code uses this syscall to check whether the given address is
+     * readable. E.g.
+     * github.com/abseil/abseil-cpp/blob/master/absl/debugging/internal/
+     * address_is_readable.cc#L85
+     * Those uses as well as our checks below don't guarantee that the given
+     * address will _remain_ readable or writable, even if they succeed now.
+     * TODO i#5255: DR can atleast pass the safe_set to the actual syscall
+     * (in cases where it is not skipped) to avoid a second read of app_set,
+     * by the kernel.
+     */
+    if (sigsetsize != sizeof(kernel_sigset_t)) {
+        if (error_code != NULL)
+            *error_code = EINVAL;
         return false;
     }
+    if (app_set != NULL && !d_r_safe_read(app_set, sizeof(safe_set), &safe_set)) {
+        if (error_code != NULL)
+            *error_code = EFAULT;
+        return false;
+    }
+    if (oset != NULL) {
+        /* Old sigset should be writable too. */
+        if (!d_r_safe_read(oset, sizeof(safe_old_set), &safe_old_set) ||
+            !safe_write_ex(oset, sizeof(safe_old_set), &safe_old_set, NULL)) {
```

Comment:
I wonder if there is some corner case like failing with EINVAL below where the kernel doesn't clobber the old set but we do here.

Reply:
The kernel implementation checks for size EINVAL first, then new sigset EFAULT, then 'how' EINVAL, and then old sigset EFAULT. In fact, the abseil code relies on the 'how' EINVAL check happening after the new sigset EFAULT check.

We currently do the two EFAULT checks together. I'll move the old sigset check to last 

https://elixir.bootlin.com/linux/latest/source/kernel/signal.c#L3162
https://elixir.bootlin.com/linux/latest/source/kernel/signal.c#L3076

Reply:
And also some more tests to verify this order of checks

Reply:
Done.